lib/src/arithmetic/compound_adder.dart

```diff
@@ -0,0 +1,148 @@
+  /// result and [sum1] sum + 1. 
+  /// The width of input [a] and [b] must be the same.
+  CompoundAdder(super.a, super.b, {super.name}) {
+    if (a.width != b.width) {
+      throw RohdHclException('inputs of a and b should have same width.');
+    }
+    addOutput('sum1', width: a.width + 1);
+  }
+}
+
+/// A trivial compound adder.
+class MockCompoundAdder extends CompoundAdder {
+
+  /// Constructs a [CompoundAdder].
+  MockCompoundAdder(
+    super.a,
+    super.b,
+    {super.name = 'mock_compound_adder'} 
+  ) {
+    sum <= a.zeroExtend(a.width + 1) + b.zeroExtend(b.width + 1);
+    sum1 <= sum + 1;
+  }
+}
+
+/// Carry-select compound adder.
+class CarrySelectCompoundAdder extends CompoundAdder {
+  
+  /// Adder ripple-carry block size computation algorithm.
+  /// Generates only one carry-select block
+  /// Return list of carry-ripple block sizes starting from 
+  /// the LSB connected one.
+  /// [adderWidth] is a whole width of adder.
+  static List<int> splitSelectAdderAlgorithmSingleBlock(int adderWidth) {
+    final splitData = <int>[adderWidth];
+    return splitData;
+  }
+
+  /// Adder ripple-carry block size computation algorithm.
+  /// Generates 4 bit carry-select blocks with 1st entry width adjusted down.
+  /// Return list of carry-ripple block sizes starting from 
+  /// the LSB connected one.
+  /// [adderWidth] is a whole width of adder.
+  static List<int> splitSelectAdderAlgorithm4Bit(int adderWidth) {
+    final blockNb = (adderWidth / 4.0).ceil();
+    final firstBlockSize = adderWidth - (blockNb - 1) * 4;
+    final splitData = <int>[firstBlockSize];
+    for (var i = 1; i < blockNb; ++i) {
+      splitData.add(4);
+    }
+    return splitData;
+  }
+
+  /// Constructs a [CarrySelectCompoundAdder].
+  CarrySelectCompoundAdder(
+    super.a,
+    super.b,
+    {super.name = 'cs_compound_adder',
+    List<int> Function(int)
+      widthGen = splitSelectAdderAlgorithmSingleBlock}
+  ) {
+    // output bits lists
+    final sumList0 = <Logic>[];
+    final sumList1 = <Logic>[];
+    // carryout of previous ripple-carry adder block
+    // for sum and sum+1
+    Logic? carry0;
+    Logic? carry1;
+    // Get size of each ripple-carry adder block
+    final adderSplit = widthGen(a.width); 
+    // 1st output bit index of each block
+    var blockStartIdx = 0;
+    for (var i = 0; i < adderSplit.length; ++i) {
+      // input width of current ripple-carry adder block
+      final blockWidth = adderSplit[i];
+      if (blockWidth <= 0) {
+        throw RohdHclException('non-positive ripple-carry adder block size.');
+      }
+      if (blockWidth + blockStartIdx > a.width) {
+        throw RohdHclException('oversized ripple-carry adders sequence.');
+      }
+      final blockA = Logic(name: 'block_${i}_a', width: blockWidth);
+      final blockB = Logic(name: 'block_${i}_b', width: blockWidth);
+      blockA <= a.slice(blockStartIdx + blockWidth - 1, blockStartIdx);
```

Comment:
a matter of preference, but just so you're aware you could use `a.getRange(blockStartIdx, blockStartIdx + blockWidth)` to save the `- 1`, if you wanted.

Reply:
updated